pkg/ccl/changefeedccl/sink_cloudstorage.go

```diff
@@ -68,11 +69,15 @@ func cloudStorageFormatTime(ts hlc.Timestamp) string {
 
 type cloudStorageSinkFile struct {
 	cloudStorageSinkKey
-	created       time.Time
-	codec         io.WriteCloser
-	rawSize       int
-	numMessages   int
-	buf           bytes.Buffer
+	created     time.Time
+	codec       io.WriteCloser
+	rawSize     int
+	numMessages int
+	buf         bytes.Buffer
+	// bufLen tracks the length of buf. Because some of the codecs write to the
+	// buffer asynchronously, we can't safely call `buf.Len()` while the codec
+	// is open.
+	bufLen        int64
```

Comment:
How awful do you think the performance would be if we wrapped the buffer with a mutex for just the pgzip writer case? I'm a little worried we might forget to update `bufLen` somewhere in the future.